src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1349,28 +1381,162 @@ private void onDecodeConnect(
 
-            final String16FW clientIdentifier = packet.clientId();
+            return reasonCode;
+        }
 
-            if (reasonCode == 0)
+        private int onDecodeConnect(
+            long traceId,
+            long authorization,
+            int progress,
+            MqttConnectFW connect)
+        {
+            final String16FW clientIdentifier = connect.clientId();
+            byte reasonCode;
+            decode:
             {
-                MqttConnectPayload payload = mqttConnectPayloadRO.decode(packet);
+                if (connected)
+                {
+                    reasonCode = PROTOCOL_ERROR;
+                    break decode;
+                }
+
+                final MqttPropertiesFW properties = connect.properties();
+
+                reasonCode = decodeConnectProperties(properties);
+
+                if (reasonCode != SUCCESS)
+                {
+                    break decode;
+                }
+
+                final MqttConnectPayload payload = mqttConnectPayloadRO.reset();
+                payload.decode(connect);
                 reasonCode = payload.reasonCode;
+
+                if (reasonCode != SUCCESS)
+                {
+                    break decode;
+                }
+
+                if (sessionExpiryInterval > 0)
+                {
+                    progress = doEncodeSession(traceId, authorization, reasonCode, progress, clientIdentifier, connect, payload);
+                }
+                else
+                {
+                    doCancelConnectTimeoutIfNecessary();
+                    doEncodeConnack(traceId, authorization, reasonCode, clientIdentifier.value());
+
+                    connected = true;
+                    keepAlive = connect.keepAlive();
+                    keepAliveTimeout = Math.round(TimeUnit.SECONDS.toMillis(keepAlive) * 1.5);
+                    doSignalKeepAliveTimeoutIfNecessary();
+                    decoder = decodePacketType;
+                    progress = connect.limit();
+                }
             }
 
-            doCancelConnectTimeoutIfNecessary();
-            doEncodeConnack(traceId, authorization, reasonCode, clientIdentifier.value());
+            if (reasonCode != SUCCESS && sessionExpiryInterval <= 0)
+            {
+                doCancelConnectTimeoutIfNecessary();
+                doEncodeConnack(traceId, authorization, reasonCode, clientIdentifier.value());
+            }
 
-            if (reasonCode == 0)
+            if (reasonCode != SUCCESS)
             {
-                connected = true;
-                keepAlive = packet.keepAlive();
-                keepAliveTimeout = Math.round(keepAlive * 1.5 * 1000);
-                doSignalKeepAliveTimeoutIfNecessary();
+                doNetworkEnd(traceId, authorization);
+                decoder = decodeIgnoreAll;
+                progress = connect.limit();
+            }
+            return progress;
+        }
+
+        private int doEncodeSession(
```

Comment:
Actually, I think we should revert this name back to `onResolveSession`, and use `doEncodeSession` for `resolveSession` below.

Not sure if it should really return `progress`, but that is minor.